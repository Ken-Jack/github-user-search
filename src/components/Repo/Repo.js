// @flow

import React from 'react';
import parse from 'date-fns/parse';
import distanceInWords from 'date-fns/distance_in_words';
import {connect} from 'react-fela';
import ForkIcon from './repo-forked.svg';
import StarIcon from './star.svg';

type Props = {
  description: string | null,
  fork: boolean,
  html_url: string,
  language: string | null,
  name: string,
  stargazers_count: number,
  pushed_at: string,
  forks_count: number,
  styles: Object,
};

function renderFooterItem(Icon: any, text: number | string, styles: Object): React$Element<*> {
  return (
    <div className={styles.Repo_footerItem}>
      <Icon
        className={styles.Repo_footerIcon}
        width={18}
        height={18}
      />
      <p>{text}</p>
    </div>
  );
}

function Repo(props: Props) {
  const {
    description,
    fork,
    html_url,
    language,
    name,
    stargazers_count,
    pushed_at,
    forks_count,
    styles,
  } = props;

  const pushedDate = distanceInWords(parse(pushed_at), new Date());

  return (
    <div>
      <p className={styles.Repo_name}>
        <a href={html_url}>{name}</a>
        {fork
          ? <span className={styles.Repo_fork}>(fork)</span>
          : null
        }
      </p>
      {description
        ? <p className={styles.Repo_desc}>{description}</p>
        : null
      }
      <div className={styles.Repo_footer}>
        {language
          ? <p className={styles.Repo_footerItem}>{language}</p>
          : null
        }
        {renderFooterItem(StarIcon, stargazers_count, styles)}
        {renderFooterItem(ForkIcon, forks_count, styles)}
        <p className={styles.Repo_pushDate}>last push {pushedDate} ago</p>
      </div>
    </div>
  );
}

const styles = {
  Repo_footerItem: () => ({
    marginRight: '12px',
    display: 'flex',
    alignItems: 'center',
  }),

  Repo_footerIcon: () => ({
    marginRight: '3px',
  }),

  Repo_footer: () => ({
    color: '#666',
    marginTop: '35px',
    fontSize: '13px',
    display: 'flex',
    flexWrap: 'wrap',
    alignItems: 'center',
  }),

  Repo_fork: () => ({
    fontSize: '12px',
    marginLeft: '5px',
    color: '#666',
  }),

  Repo_pushDate: () => ({
    marginLeft: 'auto',
  }),

  Repo_name: () => ({
    marginBottom: '10px',
    fontSize: '18px',
    display: 'flex',
    alignItems: 'baseline',
  }),

  Repo_desc: () => ({
    fontSize: '14px',
    lineHeight: '1.4',
  }),
};

export default connect(styles)(Repo);
