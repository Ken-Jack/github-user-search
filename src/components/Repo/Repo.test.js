import React from 'react';
import MockDate from 'mockdate';
import felaSnapshot from 'test-util/fela-snapshot';

import Repo from './Repo';

describe('Component: Repo', () => {

  beforeAll(() => {
    MockDate.set(new Date('2017-04-04T09:00:00Z'));
  });

  afterAll(() => {
    MockDate.reset();
  });

  it('should render all data items', () => {
    const component = felaSnapshot(
      <Repo
        description="A description"
        fork={true}
        html_url="https://github.com"
        language="JS"
        name="some repo"
        stargazers_count={10}
        pushed_at="2017-04-01T09:00:00Z"
        forks_count={2}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should not render anything for optional data items', () => {
    const component = felaSnapshot(
      <Repo
        fork={false}
        html_url="https://github.com"
        name="some repo"
        stargazers_count={10}
        pushed_at="2017-04-01T09:00:00Z"
        forks_count={2}
      />
    );
    expect(component).toMatchSnapshot();
  });

});
