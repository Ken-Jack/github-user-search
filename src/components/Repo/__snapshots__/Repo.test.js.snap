// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Repo should not render anything for optional data items 1`] = `
Object {
  "component": <div>
    <p
      className="l m b n"
    >
      <a
        href="https://github.com"
      >
        some repo
      </a>
    </p>
    <div
      className="e f g b h c"
    >
      <div
        className="a b c"
      >
        <svg />
        <p>
          10
        </p>
      </div>
      <div
        className="a b c"
      >
        <svg />
        <p>
          2
        </p>
      </div>
      <p
        className="k"
      >
        last push 
        3 days
         ago
      </p>
    </div>
  </div>,
  "styles": "
    .a{margin-right:12px}
    .b{display:flex}
    .c{align-items:center}
    .d{margin-right:3px}
    .e{color:#666}
    .f{margin-top:35px}
    .g{font-size:13px}
    .h{flex-wrap:wrap}
    .i{font-size:12px}
    .j{margin-left:5px}
    .k{margin-left:auto}
    .l{margin-bottom:10px}
    .m{font-size:18px}
    .n{align-items:baseline}
    .o{font-size:14px}
    .p{line-height:1.4}",
}
`;

exports[`Component: Repo should render all data items 1`] = `
Object {
  "component": <div>
    <p
      className="l m b n"
    >
      <a
        href="https://github.com"
      >
        some repo
      </a>
      <span
        className="i j e"
      >
        (fork)
      </span>
    </p>
    <p
      className="o p"
    >
      A description
    </p>
    <div
      className="e f g b h c"
    >
      <p
        className="a b c"
      >
        JS
      </p>
      <div
        className="a b c"
      >
        <svg />
        <p>
          10
        </p>
      </div>
      <div
        className="a b c"
      >
        <svg />
        <p>
          2
        </p>
      </div>
      <p
        className="k"
      >
        last push 
        3 days
         ago
      </p>
    </div>
  </div>,
  "styles": "
    .a{margin-right:12px}
    .b{display:flex}
    .c{align-items:center}
    .d{margin-right:3px}
    .e{color:#666}
    .f{margin-top:35px}
    .g{font-size:13px}
    .h{flex-wrap:wrap}
    .i{font-size:12px}
    .j{margin-left:5px}
    .k{margin-left:auto}
    .l{margin-bottom:10px}
    .m{font-size:18px}
    .n{align-items:baseline}
    .o{font-size:14px}
    .p{line-height:1.4}",
}
`;
